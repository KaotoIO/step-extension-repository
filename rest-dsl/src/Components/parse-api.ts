import { OpenAPI } from 'openapi-types';
import { IEndpoint } from './RestStep';
import SwaggerParser from '@apidevtools/swagger-parser';

export async function parseApiSpec(
  input: string | OpenAPI.Document
): Promise<IEndpoint[]> {
  let swaggerParser: SwaggerParser = new SwaggerParser();

  const e: Array<IEndpoint> = [];
  try {
    await swaggerParser.validate(input, { dereference: { circular: 'ignore' } });
    const paths = swaggerParser.api.paths;
    if (paths) {
      Object.keys(paths).forEach((key: string) => {
        const path = paths[key];
        if (path) {
          const operations: Map<string, OpenAPI.Operation> = new Map<string, OpenAPI.Operation>();

          Object.entries(path).forEach((method: [string, OpenAPI.Operation]) => {
            operations.set(method[0], method[1]);
          });

          operations.forEach((op: OpenAPI.Operation, verb: string) => {
            let produces: Map<string, string[]> = new Map<string, string[]>();
            let consumes: Map<string, string[]> = new Map<string, string[]>();
            let produce: Map<string, string> = new Map<string, string>();
            let consume: Map<string, string> = new Map<string, string>();
            let operation: Map<string, OpenAPI.Operation> = new Map<string, OpenAPI.Operation>();
            operation.set(verb, op);
            const producesMediaType: string[] = [];

            if (typeof op !== 'object') {
              return;
            }

            if ('produces' in op) {
              op.produces?.forEach((prod: string) => producesMediaType.push(prod));
              if (producesMediaType.length > 0) {
                produce.set(verb, producesMediaType[0]);
              }
            } else if (op.responses) {
              Object.values(op.responses).forEach((response) => {
                if (response.content) {
                  producesMediaType.push.apply(producesMediaType, (Object.keys(response.content)));
                }
              });
            }

            produces.set(verb, producesMediaType);

            if ('consumes' in op) {
              const consumesMediaTypes: string[] = [];
              consumes.set(verb, consumesMediaTypes);
              op.consumes?.forEach((con: string) => consumesMediaTypes.push(con));
              if (consumesMediaTypes.length > 0) {
                consume.set(verb, consumesMediaTypes[0]);
              }
            }

            // @ts-expect-error | The problem comes from `verb` not being recognized as a HttpMethods from OpenAPIV2, OpenAPIV3, OpenAPIV3_1
            e.push({ name: key, path: path[verb].operationId, operations: operation, produces, consumes, produce, consume });
          });
        }
      });
    }
  } catch (error) {
    console.error(error);
  }
  return e;
}
