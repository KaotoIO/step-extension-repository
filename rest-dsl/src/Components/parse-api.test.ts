import { parseApiSpec } from './parse-api';
import { readFile } from 'fs/promises';
import { resolve } from 'path';
import { parse } from 'yaml';

describe('parseApiSpec', () => {
  it('should parse an OpenAPI spec', async () => {
    const filename = resolve('src/Components/test/open-api_wso-mock.yaml');
    const openApiSpec = await readFile(filename, { encoding: 'utf8' });

    const spec = await parseApiSpec(parse(openApiSpec));

    expect(spec).toMatchSnapshot();
  });
});
