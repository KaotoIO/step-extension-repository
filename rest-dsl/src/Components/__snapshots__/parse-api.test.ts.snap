// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseApiSpec should parse an OpenAPI spec 1`] = `
[
  {
    "consume": Map {},
    "consumes": Map {},
    "name": "/token",
    "operations": Map {
      "post" => {
        "operationId": "mock1",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "string",
              },
            },
          },
          "description": "Mock request body",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "examples": {
                  "token example": {
                    "value": {
                      "access_token": "4e188f51-6ff4-3eb0-bb8b-235b652b123a",
                      "expires_in": 3600,
                      "token_type": "Bearer",
                    },
                  },
                },
              },
            },
            "description": "returns specific token",
          },
        },
        "summary": "mock of the token endpoint",
      },
    },
    "path": "mock1",
    "produce": Map {},
    "produces": Map {
      "post" => [
        "application/json",
      ],
    },
  },
  {
    "consume": Map {},
    "consumes": Map {},
    "name": "/status",
    "operations": Map {
      "get" => {
        "operationId": "getstatus",
        "responses": {
          "200": {
            "description": "Get status",
          },
        },
      },
    },
    "path": "getstatus",
    "produce": Map {},
    "produces": Map {
      "get" => [],
    },
  },
]
`;
